docs/adr/2022-12-01-021-persistency.md

```diff
@@ -0,0 +1,138 @@
+This is alos when we store them on disk that we give an id and a timestamp to each of these outputs:
+
+```haskell
+  appendToHistory history output = do
+    time <- getCurrentTime
+    timedOutput <- atomically $ do
+      seq <- nextSequenceNumber <$> readTVar history
+      let timedOutput = TimedServerOutput{output, time, seq}
+      modifyTVar' history (timedOutput :)
+      pure timedOutput
+    append timedOutput
+    pure timedOutput
+```
+
+The _append_ function uses [withBinaryFileDurable](https://hackage.haskell.org/package/unliftio-0.2.23.0/docs/UnliftIO-IO-File.html#v:withBinaryFileDurable)
+ to append data on disk. Some reliability is ensured since we only append data to the file and never change existing data.
+We may have problem if the last line of the file contains a partially writtent JSON object.
+
+Also note that, with the current implementation, we open and close the file for every server-output we want to persist.
+
+#### Drawbacks
+
+The current solution has several drawbacks. We describe some of them here.
+
+Extra disk usage with the state persistency:
+* copy of the file before overriding everything;
+* Serialization of the whole state evrytime.
+
+The state is maintained in two several files, one for server-output and one for state which may lead to incoherences in some corner case situations.
+
+When a customer send a _getUTxO_ query, for instance, the answer to this query is stored in the _server-out_ file and will then be replayed after a restart
+eventhough it wouldn't make much sense as the newly connected client didn't make any request. It could be helpful do differentiate between request answers
+and actual events which should be persisted.
+
+When a new peer is seen, a message is sent to the API client and this is stored in _server-out_. So on restart, the message will be replayed after a restart
+and the client will be informed that the peer is connected eventhough it is certainly not yet connected, just after a restart. As a counter measure, we
+also send a _Greetings_ message, later, to notify the client that the hydra node has been restarted and that it should _forget_ about some information
+received before. But may be we could have a persistent that could distinguish volatile events from persistent ones?
+
+
+### event/command sourcing as an alternative
+
+[Event sourcing](https://martinfowler.com/eaaDev/EventSourcing.html) could be an alternative to store the state of the Hydra node. In that kind of architecture,
+everytime something happens that could update the state, an event is generated, then stored on disk, then applied to the state.
+
+Some of the values could be:
+* More effecient perisitence of the state (only events are happended on disk, not the full state and no need for file copying);
+* reliability as the state is only updated after the events have been persisted on disk.
```

Comment:
This is tricky right? Persisting something to disk and being sure that it's definitely written is not easy (or impossible). But, I do agree it's _more_ reliable than writing an aggregated state. Especially if some appended events are missing, it's not as fatal.